outlines/models/llamacpp.py

```diff
@@ -1,304 +1,185 @@
-        # To make it compatible with the transformers inspired tokenizer interface
-        # we need a padding token to homogenize to token_ids tensor.
-        self.pad_token_id = -1
-
-        self.n_past = 0
-        self.n_vocab = kwargs.pop("n_vocab")
-
-        self.ctx = llama_instance.llama_new_context_with_model(model, context_params)
-
-    def forward(self, input_ids: torch.LongTensor, *_):
-        """Compute a forward pass through the llama_cpp model."""
-        if input_ids.ndim == 2:
-            seq_tensor = input_ids[:, self.n_past :]
-        elif input_ids.ndim == 1:
-            seq_tensor = input_ids.view(1, -1)[:, self.n_past :]
-        else:
-            raise Exception("Only one and two dimensional inputs allowed.")
-
-        tokens_total = torch.numel(seq_tensor[seq_tensor != self.pad_token_id])
-        batch = self.llama_instance.llama_batch_init(tokens_total, 0, 1)
-
-        seq_token_ids = []
-        for seq_idx, seq in enumerate(seq_tensor):
-            for token_pos, token_id in enumerate(seq):
-                if token_id == self.pad_token_id:
-                    break
-                batch.token[batch.n_tokens] = token_id.item()
-                batch.pos[batch.n_tokens] = token_pos
-                batch.seq_id[batch.n_tokens][0] = seq_idx
-                batch.n_seq_id[batch.n_tokens] = 1
-                batch.logits[batch.n_tokens] = False
-
-                batch.n_tokens += 1
-                self.n_past += 1
-
-            batch.logits[batch.n_tokens - 1] = True
-            seq_token_ids.append(batch.n_tokens - 1)
+        from llama_cpp import Llama
 
-        if self.llama_instance.llama_decode(self.ctx, batch) != 0:
-            print("Error decoding")
+        self.model = Llama(model_path, **kwargs)
+        self.logits_processor = logits_processor
+        self.tokenizer = LlamaCppTokenizer(self.model)
 
-        all_logits = []
-        for seq_token in seq_token_ids:
-            logits = self.llama_instance.llama_get_logits_ith(self.ctx, seq_token)
-            logits_list = (ctypes.c_float * self.n_vocab)(
-                *[logits[token_id] for token_id in range(self.n_vocab)]
-            )
-            logits_tensor = torch.tensor(logits_list)
-            all_logits.append(logits_tensor)
+    def __call__(
+        self,
+        prompts: Union[str, List[str]],
+        max_tokens: Optional[int] = None,
+        stop_at: Optional[Union[str, List[str]]] = None,
+        rng: Optional[torch.Generator] = None,
+        **model_kwargs,
+    ) -> Union[str, List[str]]:
+        from llama_cpp import LogitsProcessorList
+
+        if isinstance(prompts, str):
+            prompts = [prompts]
+
+        if rng is None:
+            rng = torch.Generator(device="cpu")
+            rng.seed()
+
+        results = []
+        for prompt in prompts:
+            processors = []
+            if self.logits_processor is not None:
+                processors = [copy.copy(self.logits_processor)]
```

Comment:
Won't the logits processors share a `CFGFSM` if we shallow copy?

Reply:
True, logits processors should have a `copy` method.

Reply:
We should probably open an issue for this.